components/SerialServer/src/serial.c

```diff
@@ -310,44 +279,56 @@ static void internal_putchar(int b, int c)
     error = serial_unlock();
 }
 
-static void internal_raw_putchar(int id, int c)
+static int serial_notify_recv(int client)
 {
-    getchar_client_t *client = &getchar_clients[id];
-    uint32_t next_tail = (client->buf->tail + 1) % sizeof(client->buf->buf);
-    if (next_tail == client->buf->head) {
-        /* full */
-        return;
+    volatile serial_shmem_t *buffer = clients[client].recv_buf;
+    while (buffer->head != buffer->tail) {
+        char ch = buffer->buf[buffer->head];
+        if (ch == '\n') {
+            internal_putchar(client, '\r');
+        }
+        internal_putchar(client, ch);
+        buffer->head = (buffer->head + 1) % BUFSIZE;
     }
-    uint32_t last_read_head = client->buf->head;
-    client->buf->buf[client->buf->tail] = (uint8_t)c;
-    /* no synchronize in here as we assume TSO */
-    client->buf->tail = next_tail;
-    __sync_synchronize();
-    if (last_read_head != client->last_head) {
-        getchar_emit(client->client_id);
-        client->last_head = last_read_head;
+}
+
+/* This is called whenever one of our virtqueues get notified */
+void serial_wait_callback(void)
+{
+    /* Walk through all recv buffers and poll for data */
+    for (int i = 0; i < num_clients; i++) {
+        if (clients[i].recv_buf->head != clients[i].recv_buf->tail) {
+            serial_notify_recv(i);
+        }
     }
 }
 
-static int statemachine = 1;
+static void internal_raw_putchar(int id, int c)
+{
+    volatile serial_shmem_t *putchar_buf = clients[id].send_buf;
+    putchar_buf->buf[putchar_buf->tail] = c;
+    putchar_buf->tail = (putchar_buf->tail + 1) % BUFSIZE;
+    clients[id].notify();
+}
 
 static void give_client_char(uint8_t c)
 {
     if (active_client >= 0) {
         internal_raw_putchar(active_client, c);
     } else if (active_client == -1) {
-        for (int i = 0; i < MAX_CLIENTS * 2; i++) {
+        for (int i = 0; i < MAX_CLIENTS; i++) {
```

Comment:
Why using `MAX_CLIENTS` instead of `num_clients`? Or does `MAX_CLIENTS` mean something different?